### DISCUSSION TITLE
replica cluster reports "Cluster is ready" when replication is broken

### DISCUSSION THREAD

**Discussion by cleboo** [cloudnative-pg/cloudnative-pg#1285] [Q&A]
Hi!

Thanks in advance for your help!

## The issue
I am using cnpg to set up a replica cluster that is connected to its primary using streaming_replication. When a connection to the primary is not possible (due to e.g. a firewall misconfiguration), the replication fails with the following error message:
```json
{
  "level": "info",
  "ts": 1664963681.2284603,
  "logger": "postgres",
  "msg": "record",
  "logging_pod": "db-1",
  "record": {
    "log_time": "2022-10-05 09:54:41.228 UTC",
    "process_id": "282085",
    "session_id": "633d5461.44de5",
    "session_line_num": "1",
    "session_start_time": "2022-10-05 09:54:41 UTC",
    "transaction_id": "0",
    "error_severity": "FATAL",
    "sql_state_code": "08006",
    "message": "could not connect to the primary server: connection to server at \"XXX\" (xxx.xxx.xxx.xxx), port XXXXX failed: server closed the connection unexpectedly\n\tThis probably means the server terminated abnormally\n\tbefore or while processing the request.",
    "backend_type": "walreceiver",
    "query_id": "0"
  }
}
```

At the same time the cluster condition is reported as "Cluster is Ready":
```
$ k get cluster db -o json | jq -r '.status.conditions'
[
  {
    "lastTransitionTime": "2022-10-04T13:09:52Z",
    "message": "Cluster is Ready",
    "reason": "ClusterIsReady",
    "status": "True",
    "type": "Ready"
  }
]
```

## What should happen?
This should trigger an error instead, reporting that the replication from the primary is broken.

**Comment by sxd** [comment]
Hi @cleboo 

Can you provide the output of the `kubectl cnpg status` command for this cluster?

**Comment by cleboo** [comment]
Hi @sxd 

Thanks for looking into this. Here the output for the replica cluster that can not connect to its primary.  I attached the manifest as well (with some info redacted):

```
Cluster Summary
Name:              db
Namespace:         default
PostgreSQL Image:  ghcr.io/cloudnative-pg/postgresql:14.5
Primary instance:  db-1
Status:            Cluster in healthy state
Instances:         2
Ready instances:   2

Certificates Status
Certificate Name        Expiration Date                Days Left Until Expiration
----------------        ---------------                --------------------------
db-client-ca         2023-11-14 15:05:57 +0000 UTC  336.09
db-replication-cert  2023-11-14 15:05:57 +0000 UTC  336.09
db-server            2023-03-06 16:03:05 +0000 UTC  83.13
db-server-ca         2023-11-14 15:05:57 +0000 UTC  336.09

Continuous Backup status
Not configured

Streaming Replication status
Primary instance not found

Instances status
Name     Database Size  Current LSN  Replication role      Status  QoS         Manager Version  Node
----     -------------  -----------  ----------------      ------  ---         ---------------  ----
db-1  11 GB          5C1/F000000  Standby (file based)  OK      Guaranteed  1.17.0           node05
db-2  11 GB          5C1/F000000  Unknown               OK      Guaranteed  1.17.0           node06
```

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  generation: 1
  labels:
    app.kubernetes.io/instance: app
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: base
    app.kubernetes.io/version: 0.1.0
    helm.sh/chart: base-0.1.0
  name: db
  namespace: default
  resourceVersion: "32857747"
  uid: 63a557ad-b8a8-4f0e-a272-328a432a0089
spec:
  affinity:
    podAntiAffinityType: preferred
    topologyKey: ""
  bootstrap:
    pg_basebackup:
      database: ""
      owner: ""
      source: db
  certificates:
    clientCASecret: db-client-ca
    replicationTLSSecret: db-cert
    serverAltDNSNames:
    - <REDACTED>
    serverCASecret: db-ca
  enableSuperuserAccess: true
  externalClusters:
  - connectionParameters:
      dbname: postgres
      host: <REDACTED>
      port: <REDACTED>
      sslmode: verify-full
      user: streaming_replica
    name: db
    sslCert:
      key: tls.crt
      name: db-replication-cert
    sslKey:
      key: tls.key
      name: db-replication-cert
    sslRootCert:
      key: ca.crt
      name: db-server-ca
  imageName: ghcr.io/cloudnative-pg/postgresql:14.5
  instances: 2
  logLevel: info
  maxSyncReplicas: 1
  minSyncReplicas: 1
  monitoring:
    customQueriesConfigMap:
    - key: queries
      name: cnpg-default-monitoring
    disableDefaultQueries: false
    enablePodMonitor: false
  postgresGID: 26
  postgresUID: 26
  postgresql:
    parameters:
      archive_mode: always
      archive_timeout: 5min
      dynamic_shared_memory_type: posix
      log_destination: csvlog
      log_directory: /controller/log
      log_filename: postgres
      log_rotation_age: "0"
      log_rotation_size: "0"
      log_truncate_on_rotation: "false"
      logging_collector: "on"
      max_parallel_workers: "32"
      max_replication_slots: "32"
      max_slot_wal_keep_size: 20GB
      max_worker_processes: "32"
      shared_buffers: 1GB
      shared_memory_type: mmap
      shared_preload_libraries: ""
      wal_keep_size: 512MB
      wal_receiver_timeout: 5s
      wal_sender_timeout: 5s
    syncReplicaElectionConstraint:
      enabled: false
  primaryUpdateMethod: switchover
  primaryUpdateStrategy: unsupervised
  replica:
    enabled: true
    source: db
  resources:
    limits:
      cpu: "1"
      memory: 4Gi
    requests:
      cpu: "1"
      memory: 4Gi
  startDelay: 30
  stopDelay: 30
  storage:
    resizeInUseVolumes: true
    size: 50Gi
  superuserSecret:
    name: db-secret
  switchoverDelay: 40000000
```

cloutdative-pg version in `v1.17.0`.

**Comment by sxd** [comment]
Hi @cleboo 

Sorry for the late reply, can you confirm if you can still see the same issue with version 1.18.1?
Also, this looks like an issue with the storage or networking, can you check the logs looking for some information that may look like "cannot stream WAL file" or something like that?

Will turn this into a discussion, since I don't see an issue and this looks more like a configuration problem or either a cluster issue.

Best Regards,

**Reply by cleboo** [reply]
Hi @sxd!

> Sorry for the late reply, can you confirm if you can still see the same issue with version 1.18.1?

Sure, I'll test with the new version.

> Sorry for the late reply, can you confirm if you can still see the same issue with version 1.18.1?
> Also, this looks like an issue with the storage or networking, can you check the logs looking for some information that may look like "cannot stream WAL file" or something like that?

True, the fact that the replication is not working is a configuration issue - I actually blocked the network connection on the firewall on purpose to demonstrate. 

The issue I want to report is that the status of the Cluster on the replica side is still reported as "okay" even though the replication is broken. In order to monitor the replication it would be great, if the replica side shows a "broken" state if it can detect that the connection is not working.

Best,

**Reply by sxd** [reply]
Hi @cleboo 

The problem with that approach it's that the instance itself can't report the status of cluster to anyone, isn't that the case? because if that's the case the instance will not update the cluster status, in that case, no one will report that failure, what you can see here it's that it's file based, meaning that it will try to retrieve the WAL files from the streaming or from the primary, which it's not happening because that connection it's also blocked, but for PostgreSQL that it's just a different state.

With that being said, how you think it can inform that the cluster it's broken if there's no network connection? It will be great to hear some more options

Regards!

**Comment by cleboo** [comment]
Hi,

ah sorry - I think we are talking about different things. To clarify: I am running a replica cnpg-cluster in a totally separate k8s instance (replica-k8s) that is replicated from a primary cnpg-cluster (in a primary k8s-cluster). Basically the setup shown in the manual here: https://cloudnative-pg.io/documentation/1.17/replica_cluster/

For demonstration I cut replica-k8s off from primary-k8s by blocking the network connection on the firewall between the two clusters. Now, the replica cluster can not fetch the WAL files from the primary cluster.

What I was looking for is that the replica cluster reports that it is degraded to the **replica**-k8s cluster (which it can still connect to) so that the failure to fetch WAL files shows up in the `.status.conditions` array of the cluster in replica-k8s. That way the failure can be displayed in `kubectl cnpg status` and in other tools when the replica k8s cluster is queried.

Appreciate, you're taking the time!
